File: bindings/python/src/kortecx/defaults.py
"""Zero-config client resolution (Batch V2).

``import kortecx as kx; kx.run(...)`` uses a lazily-built, process-wide default client
so the simplest path needs no constructor. Config order (first wins):

1. explicit kwargs,
2. environment — ``KX_ENDPOINT`` / ``KX_TOKEN`` / ``KX_DEFAULT_MODEL``,
3. ``~/.kortecx/config.toml`` (a ``[client]`` table),
4. the conventional defaults (loopback endpoint, no token, server-bound model).

The explicit :class:`~kortecx.client.KxClient` / :class:`~kortecx.client.AsyncKxClient`
stay available for full control; this module is the convenience layer on top.
"""

from __future__ import annotations

import os
from pathlib import Path
from typing import Optional, Union

from .client import DEFAULT_ENDPOINT, KxClient

#: What the gateway accepts as recipe args (mirrors the client's ``ArgsType``).
ArgsType = Union[dict, bytes, bytearray, str]

#: The process-wide default client (lazily built on first use).
_DEFAULT_CLIENT: Optional[KxClient] = None
_CONFIG_PATH = Path.home() / ".kortecx" / "config.toml"


def _load_config() -> dict:
    """Best-effort read of the ``[client]`` table from ``~/.kortecx/config.toml``.
    A missing file / no ``tomllib`` (Python < 3.11) / a parse error ⇒ ``{}`` (env +
    defaults still apply). Never raises."""
    try:
        import tomllib  # type: ignore[import-not-found]  # Python 3.11+
    except ModuleNotFoundError:
        return {}
    try:
        with _CONFIG_PATH.open("rb") as fh:
            data = tomllib.load(fh)
    except (OSError, ValueError):
        return {}
    section = data.get("client")
    return section if isinstance(section, dict) else {}


def resolve_endpoint(explicit: Optional[str] = None) -> str:
    """The gateway endpoint: explicit → ``KX_ENDPOINT`` → config → the loopback default."""
    cfg = _load_config()
    return explicit or os.environ.get("KX_ENDPOINT") or cfg.get("endpoint") or DEFAULT_ENDPOINT


def resolve_default_model(explicit: str = "") -> str:
    """The default model id: explicit → ``KX_DEFAULT_MODEL`` → config → ``""`` (the
    server binds the served model, SN-8)."""
    cfg = _load_config()
    return explicit or os.environ.get("KX_DEFAULT_MODEL") or cfg.get("default_model") or ""


def make_client(
    endpoint: Optional[str] = None,
    *,
    token: Optional[str] = None,
    default_model: str = "",
) -> KxClient:
    """Build a :class:`~kortecx.client.KxClient` from explicit args + env +
    ``~/.kortecx/config.toml`` + the defaults."""
    cfg = _load_config()
    tok = token or os.environ.get("KX_TOKEN") or cfg.get("token")
    return KxClient(
        resolve_endpoint(endpoint),
        token=tok,
        default_model=resolve_default_model(default_model),
    )


def default_client() -> KxClient:
    """The lazily-built, process-wide default client used by the module-level
    ``kx.run`` / ``kx.invoke`` and the :class:`~kortecx.flow.Flow` /
    :class:`~kortecx.agent.Agent` terminals.

    NOTE (async / threads): this is a SINGLE shared client (one gRPC channel) — fine
    for scripts and sync use; for concurrent or async work, construct explicit
    :class:`~kortecx.client.KxClient` / :class:`~kortecx.client.AsyncKxClient`
    instances rather than relying on this singleton."""
    global _DEFAULT_CLIENT
    if _DEFAULT_CLIENT is None:
        _DEFAULT_CLIENT = make_client()
    return _DEFAULT_CLIENT


def set_default_client(client: Optional[KxClient]) -> None:
    """Override (or clear, with ``None``) the process-wide default client — for tests
    or a custom endpoint/token without threading a client through every call."""
    global _DEFAULT_CLIENT
    _DEFAULT_CLIENT = client


def run(target: object, *, wait: bool = True, timeout: float = 120.0, **agent_kwargs):
    """Module-level convenience — run a :class:`~kortecx.flow.Flow`, a
    :class:`~kortecx.chains.Chain`, or a bare prompt (a one-line agent) via the default
    client. ``agent_kwargs`` (``tools=`` / ``reasoning=`` / …) apply only to the prompt
    form."""
    from .chains import Chain
    from .flow import Flow
    from .flow import flow as _flow

    kx = default_client()
    if isinstance(target, Flow):
        return target.run(wait=wait, timeout=timeout, client=kx)
    if isinstance(target, Chain):
        return kx.run_chain(target, wait=wait, timeout=timeout)
    if isinstance(target, str):
        return _flow().agent(target, **agent_kwargs).run(wait=wait, timeout=timeout, client=kx)
    raise TypeError(f"kx.run() accepts a Flow, Chain, or prompt str, got {type(target).__name__}")


def invoke(handle: str, args: "ArgsType", **kwargs):
    """Module-level convenience — invoke a published recipe via the default client."""
    return default_client().invoke(handle, args, **kwargs)

File: bindings/typescript/src/defaults.ts
/**
 * Zero-config client resolution for Node (V2a g1) — the TS mirror of Python's
 * `kortecx.defaults`.
 *
 * ```ts
 * import { run, flow } from "@kortecx/sdk";
 * const out = await run(flow().agent("Summarize the README"));
 * console.log((out as Result).text);
 * ```
 *
 * A lazily-built, process-wide default client so the simplest path needs no
 * constructor. Config order (first wins):
 *
 * 1. explicit args,
 * 2. environment — `KX_ENDPOINT` / `KX_TOKEN` / `KX_DEFAULT_MODEL`,
 * 3. `~/.kortecx/config.toml` (a `[client]` table),
 * 4. the conventional defaults (loopback endpoint, no token, server-bound model).
 *
 * NODE-ONLY: reads `node:fs`/`node:os` and builds the gRPC {@link KxClient}. The
 * `web` and `chains` entrypoints do NOT import this module (explicit-client by
 * design), so it never reaches a browser bundle. On load it installs the lazy
 * factory the {@link Flow}/{@link Agent} terminals use for a zero-config client.
 */

import { readFileSync } from "node:fs";
import { homedir } from "node:os";
import { join } from "node:path";
import type { Chain } from "./chains.js";
import type { InvokeOptions } from "./client.js";
import {
  getDefaultClient,
  setDefaultClient as registrySet,
  setDefaultClientFactory,
} from "./default-client.js";
import { KxUsage } from "./errors.js";
import { type AgentStepOptions, Flow, flow } from "./flow.js";
import { KxClient } from "./node.js";
import type { Result, Run } from "./run.js";
import { type Args, DEFAULT_ENDPOINT } from "./transport.js";

interface ClientConfig {
  endpoint?: string;
  token?: string;
  default_model?: string;
}

/**
 * Best-effort read of the `[client]` table from `~/.kortecx/config.toml`. A missing
 * file / read error ⇒ `{}` (env + defaults still apply). Never throws. A minimal,
 * dependency-free reader: only `key = "value"` lines inside the `[client]` table are
 * read (the three keys the SDK uses) — not a general TOML parser.
 */
function loadConfig(): ClientConfig {
  let text: string;
  try {
    text = readFileSync(join(homedir(), ".kortecx", "config.toml"), "utf-8");
  } catch {
    return {};
  }
  const out: ClientConfig = {};
  let inClient = false;
  for (const raw of text.split(/\r?\n/)) {
    const line = raw.trim();
    if (line === "" || line.startsWith("#")) continue;
    if (line.startsWith("[")) {
      inClient = line === "[client]";
      continue;
    }
    if (!inClient) continue;
    const eq = line.indexOf("=");
    if (eq < 0) continue;
    const key = line.slice(0, eq).trim();
    let val = line.slice(eq + 1).trim();
    if ((val.startsWith('"') && val.endsWith('"')) || (val.startsWith("'") && val.endsWith("'"))) {
      val = val.slice(1, -1);
    }
    if (key === "endpoint") out.endpoint = val;
    else if (key === "token") out.token = val;
    else if (key === "default_model") out.default_model = val;
  }
  return out;
}

function resolveEndpoint(explicit?: string): string {
  return explicit || process.env.KX_ENDPOINT || loadConfig().endpoint || DEFAULT_ENDPOINT;
}

function resolveDefaultModel(explicit?: string): string {
  return explicit || process.env.KX_DEFAULT_MODEL || loadConfig().default_model || "";
}

/** Build a {@link KxClient} from explicit args + env + `~/.kortecx/config.toml` + defaults. */
export function makeClient(
  opts: { endpoint?: string; token?: string; defaultModel?: string } = {},
): KxClient {
  const cfg = loadConfig();
  const token = opts.token ?? process.env.KX_TOKEN ?? cfg.token;
  return new KxClient(resolveEndpoint(opts.endpoint), {
    token,
    defaultModel: resolveDefaultModel(opts.defaultModel),
  });
}

// Install the lazy factory the bundle-safe registry uses so `flow().run()` /
// `agent().run()` / `run(...)` work with no explicit client in Node. The guard
// mirrors Python's `default_client()`: a top-level run inside the `_toolserver`
// re-import context would recurse, so refuse it loudly.
setDefaultClientFactory(() => {
  if (process.env.KX_TOOLSERVE) {
    throw new KxUsage(
      "cannot start a run while serving localTool functions (the _toolserver re-import " +
        "context); guard your run()/invoke() calls so they do not re-execute on import",
    );
  }
  return makeClient();
});

/**
 * The lazily-built, process-wide default Node client used by the module-level
 * {@link run} / {@link invoke} and the {@link Flow} / {@link Agent} terminals.
 *
 * NOTE (concurrency): this is a SINGLE shared client (one transport) — fine for
 * scripts and sync use; for concurrent work construct explicit {@link KxClient}
 * instances rather than relying on this singleton.
 */
export function defaultClient(): KxClient {
  return getDefaultClient() as KxClient;
}

/** Override (or clear, with `null`) the process-wide default client — for tests or a
 * custom endpoint/token without threading a client through every call. */
export function setDefaultClient(client: KxClient | null): void {
  registrySet(client ?? undefined);
}

/** Module-level convenience — run a {@link Flow}, a {@link Chain}, or a bare prompt (a
 * one-line agent) via the default client. `opts` agent fields (`tools` / `reasoning` /
 * …) apply only to the prompt form. */
export function run(
  target: Flow | Chain | string,
  opts: { wait?: boolean; timeoutMs?: number } & AgentStepOptions = {},
): Promise<Run | Result> {
  const kx = defaultClient();
  if (target instanceof Flow) {
    return target.run({ wait: opts.wait, timeoutMs: opts.timeoutMs, client: kx });
  }
  if (typeof target === "string") {
    const { wait, timeoutMs, ...agentOpts } = opts;
    return flow().agent(target, agentOpts).run({ wait, timeoutMs, client: kx });
  }
  return kx.runChain(target, { wait: opts.wait ?? true, timeoutMs: opts.timeoutMs });
}

/** Module-level convenience — invoke a published recipe via the default client. */
export function invoke(handle: string, args: Args, opts?: InvokeOptions): Promise<Run | Result> {
  return defaultClient().invoke(handle, args, opts);
}

/** Module-level convenience — a grounded chat via the default client (the TS mirror of
 * Python's `kx.chat`). Pass `dataset` (+ optional `k`) to auto-RAG the answer over a
 * dataset's retrieved text; returns the answer text. */
export function chat(
  prompt: string,
  opts: { dataset?: string; k?: number; timeoutMs?: number } = {},
): Promise<string> {
  return defaultClient().chat(prompt, opts);
}
